### PR TITLE
Reuse resource when reparented for some extensions

### LINKED ISSUE #6987: amp-list being render twice from google cache
Please walk through the following steps for reporting an issue about AMP:

## Short description of your issue:

Just notice recently, our amp-list being render twice via google cache

## How do we reproduce the issue?

1. Access https://www.google.com/amp/s/bookmate.app.link/foodcombining
1. Notice that there are *two* amp-list xhr fired and render twice amp list items.
1. Accessing directly via the link works fine https://bookmate.app.link/foodcombining

The `<template>` tag from the markup looks fine.

Also notice that if passing #development=1, we can see the same issues as well

## What browsers are affected?

All browsers? Some specific browser? What device type?
All browsers

## Which AMP version is affected?
`Powered by AMP ⚡ HTML – Version 1483661191373 `

The google cache issue happens just recently while the debug mode is always there.

### PR DESCRIPTION
**Pull request by dvoytenko**
Closes #6987.

Notice that `reconstructWhenReparented` is the current behavior. This PR enables a non-reconstructing mode. Follow up PRs will update this flag for other AMP elements.

### CODE REVIEW

File: src/base-element.js [hunk 1]
```diff
@@ -447,6 +447,16 @@ export class BaseElement {
   }
 
   /**
+   * Whether the element needs to be reconstructed after it has been
+   * re-parented. Many elements cannot survive fully the reparenting and
+   * are better to be reconstructed from scratch.
```
Comment (cramforce) [inline src/base-element.js]: Give example when this is not the case.

Comment (dvoytenko) [inline src/base-element.js]: Done

File: src/service/resources-impl.js [hunk 2]
```diff
@@ -362,11 +362,21 @@ export class Resources {
       this.viewport_.ensureReadyForElements();
     }
 
-    // Create and add the resource.
-    const resource = new Resource((++this.resourceIdCounter_), element, this);
+    // First check if the resource is being reparented and if it requires
+    // reconstruction. Only already built elements are eligible.
+    let resource = Resource.forElementOptional(element);
+    if (resource &&
+        resource.getState() != ResourceState.NOT_BUILT &&
```
Comment (jridgewell) [inline src/service/resources-impl.js]: If the element has already been built, we'll end up creating a new Resource for it. Should this be inside another conditional?

```js
if (resource && !reconstructWhenReparented) {
  if (!built) {
    //...
  }
} else {
  // new Resource
}

Comment (dvoytenko) [inline src/service/resources-impl.js]: So, I'm kind of leading this from a current state, which is `reconstructWhenReparented  = true` for all cases. The `!built` case is ultimately doesn't matter - there's no state so we are not really winning or loosing anything with that. Also, `Resource` class already knows how to work correctly with "built" elements (see https://github.com/ampproject/amphtml/blob/master/src/service/resource.js#L147).

Comment (dvoytenko) [inline src/service/resources-impl.js]: I'm gonna submit. There will likely be some changes to this part of going forward.

**Pull request merged**